compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -441,21 +458,12 @@ macro_rules! define_queries {
                     }
                 }
 
-                fn try_load_from_on_disk_cache(tcx: TyCtxt<'_>, dep_node: DepNode) {
-                    debug_assert!(tcx.dep_graph.is_green(&dep_node));
-
-                    let key = recover(tcx, dep_node).unwrap_or_else(|| panic!("Failed to recover key for {:?} with hash {}", dep_node, dep_node.hash));
-                    if queries::$name::cache_on_disk(tcx, &key) {
-                        let _ = tcx.$name(key);
-                    }
-                }
-
                 DepKindStruct {
                     is_anon,
                     is_eval_always,
                     fingerprint_style,
                     force_from_dep_node: Some(force_from_dep_node),
-                    try_load_from_on_disk_cache: Some(try_load_from_on_disk_cache),
+                    try_load_from_on_disk_cache: Some(|tcx, key| $crate::plumbing::try_load_from_on_disk_cache(tcx, key, recover, queries::$name::cache_on_disk, TyCtxt::$name)),
```

Comment:
It's a bit of a pity that we have 2 monomorphised symbols here: the closure and the generic function.

Reply:
yeah :/ but I couldn't get it to work with just the generic function.

<details>

```
error[E0308]: mismatched types
    --> compiler/rustc_query_impl/src/plumbing.rs:467:47
     |
324  | /  macro_rules! define_queries {
325  | |      (
326  | |       $($(#[$attr:meta])*
327  | |          [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...    |
467  | |                      force_from_dep_node: Some($crate::plumbing::force_from_dep_node::<queries::$name<'_>>),
     | |                                           ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
     | |                                           |
     | |                                           arguments to this enum variant are incorrect
...    |
478  | |      }
479  | |  }
     | |__- in this expansion of `define_queries!` (#2)
     = note: expected fn pointer `for<'r> fn(rustc_middle::ty::TyCtxt<'r>, DepNode<_>) -> _`
                   found fn item `fn(rustc_middle::ty::TyCtxt<'_>, DepNode<_>) -> _ {force_from_dep_node::<'_, queries::typeck_item_bodies<'_>>}`
```

</details>